File: packages/client/src/lib/lyricOffset.ts
import type { Track } from '@music-together/shared'

/**
 * Bilibili videos can contain a custom intro/outro, so their lyric adjustment
 * must be scoped to the video instead of changing the matched song globally.
 */
export function getLyricOffsetKey(track: Track | null): string | null {
  if (!track?.lyricId) return null

  const lyricSource = track.metadataSource ?? track.source
  if (track.source === 'bilibili') {
    return `bilibili:${track.urlId}:${lyricSource}:${track.lyricId}`
  }
  return `${lyricSource}:${track.lyricId}`
}

File: packages/android-client/app/src/main/java/io/github/yueby/musictogether/lyrics/LyricOffset.kt
package io.github.yueby.musictogether.lyrics

import io.github.yueby.musictogether.model.Track

fun lyricOffsetKey(track: Track?): String? {
    val value = track ?: return null
    val lyricId = value.lyricId?.takeIf { it.isNotBlank() } ?: return null
    val lyricSource = value.metadataSource ?: value.source
    return if (value.source == "bilibili") {
        "bilibili:${value.urlId}:$lyricSource:$lyricId"
    } else {
        "$lyricSource:$lyricId"
    }
}
